R/if2.R

```diff
@@ -0,0 +1,279 @@
+      iterations <- private$control$iterations
+      cooling_target <- private$control$cooling_target
+      pars_sd <- private$pars_sd
+
+      pars_matrix <- private$pars$walk_initialise(n_par_sets,
+                                                  pars_sd)
+      n_pars <- nrow(pars_matrix)
+
+      model <- private$model$new(pars = private$pars$model(pars_matrix),
+                                 step = steps[[1L]],
+                                 n_particles = NULL, n_threads = n_threads,
+                                 seed = seed, pars_multi = TRUE)
+      if (!is.null(private$index)) {
+        model$set_index(private$index(mod$info())$run)
+      }
+
+      log_likelihood <- rep(0, iterations)
+      if_pars <- array(NA_real_, c(n_pars, n_par_sets, iterations))
+      alpha_cool <- cooling_target ^ (1 / iterations)
+
+      p <- pmcmc_progress(iterations, private$control$progress)
+
+      for (m in seq_len(iterations)) {
+        p()
+        model$reset(pars = private$pars$model(pars_matrix), steps[[1L]])
+        ## TODO: refactor internal loop into function
+        for (t in seq_len(n_steps)) {
+          step_end <- steps[t, 2L]
+          state <- model$run(step_end)
+
+          log_weights <- private$compare(state, data_split[[t]],
+                                         private$compare_pars)
+          log_weights <- log_weights + private$pars$prior(pars_matrix)
+
+          if (!is.null(log_weights)) {
+            weights <- scale_log_weights(log_weights)
+            log_likelihood[m] <- log_likelihood[m] + weights$average
+            if (log_likelihood[m] == -Inf) {
+              break
+            }
+
+            kappa <- particle_resample(weights$weights)
+            model$reorder(kappa)
+            pars_matrix <- private$pars$walk(pars_matrix[, kappa],
+                                             pars_sd)
+            model$set_pars(private$pars$model(pars_matrix))
+          }
+        }
+        pars_sd <- pars_sd * alpha_cool
+        pars_final <- pars_matrix
+        pars_matrix <- private$pars$walk(pars_final, pars_sd)
+        if_pars[, , m] <- pars_final
+      }
+      # outputs
+      # pars dimensions are: n_pars, n_par_sets, iterations
+      # ll dimensions are: iterations, n_par_sets
+      private$ll <- log_likelihood
+      private$if_pars <- if_pars
+    },
+
+    ##' @description Return the log-likelihood at each iteration
+    log_likelihood = function() {
+      if (is.null(private$ll)) {
+        stop("IF2 must be run first")
+      }
+      private$ll
+    },
+
+    ##' @description Return the set of parameters at each iteration
+    pars_series = function() {
+      if (is.null(private$ll)) {
+        stop("IF2 must be run first")
+      }
+      private$if_pars
+    },
+
+    ##' @description Plot results
+    ##'
+    ##' #param what Which parameter to plot. Any of `pars$names()` or `ll`
+    ##' for the log-likelihood
+    plot = function(what = "ll") {
```

Comment:
Generally, we've avoided plot methods, and I'd like to keep it that way because they're tarpits and hard to test